### ISSUE TITLE
test06 fails on Fedora 29

### ISSUE THREAD

**Issue by Dridi** [varnish/hitch#292]
This commit exhibits the problem better on my system, for some reason I haven't investigated yet the file is not created by openssl: a777d6461f35130a8602342d23c606eb092f8279.

**Comment by Dridi** [comment]
Could be related to #288 but my error is different since it fails one step earlier.

**Comment by Dridi** [comment]
I did some digging today to look at the implementation of `openssl s_client -sess_out` and gave up on pulling the thread (because dependencies grow in a quadratic fashion). I don't have time to review too much openssl code right now.

So I tried to run the test case with `strace` to follow the life of the `sess_ticket.txt` file that appears to be missing when the first `s_client` terminates without error. The problem being that running the test with `strace` it passes consistently, otherwise it fails consistently.

I love heisenbugs.

**Comment by Dridi** [comment]
If anyone else has an idea for how I could trace this problem, any alternative to `strace`, you're welcome.

I'm blanking here.

**Comment by Dridi** [comment]
It still happens on Fedora 30, but sometimes the test passes. Also it looks like this is happening on Ubuntu 19.04 too: https://github.com/varnish/hitch/pull/304#issuecomment-491215326.

**Comment by l0stman** [comment]
Hi,
I just stumbled onto this problem on Ubuntu 18.04 with OpenSSL version 1.1.1. After digging around a little bit, here's what I discovered:

- `s_client` will negotiate the highest mutually supported protocol version. It corresponds to TLS 1.3 in my case.

- Session resumption through session tickets has been superseded by 0-RTT in TLS 1.3.

Putting this together, I was able to pass the test by passing the `-no_tls1_3` option to each `s_client` command as follows:
```sh
s_client -no_tls1_3 -sess_out sess_ticket.txt >out.dump
run_cmd test -f sess_ticket.txt
s_client -no_tls1_3 -sess_in  sess_ticket.txt >in.dump
```
That being said, I have no clue why running them through `strace` would help pass the test.

**Comment by Dridi** [comment]
I never put 2 and 2 together because sometimes (and always when I trace the process) I end up with a reused TLS 1.3 session:

> Reused, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384

@daghf maybe the way we check resumption happened is not solid enough?

**Comment by Dridi** [comment]
Actually, session resumption works with TLS 1.3, but differently. I think we have a timing issue explaining why it sometimes passes for me, and always passes with any form of tracing as tracing slows things down noticeably.

I looked at how curl deals with it since we use a curl client here:

https://github.com/curl/curl/commit/549310e907e82e44c59548351d4c6ac4aaada114

Considering that session resumption works differently for TLS 1.3 we might want to keep the test case modified as suggested by @l0stman, and then add a TLS 1.3-only check. I suspect we don't properly wait for the key exchange during 1.3 resumption.

**Comment by l0stman** [comment]
You're right.  It kind of makes sense now.  Quoting from [here](https://wiki.openssl.org/index.php/TLS1.3#Sessions):

> In TLSv1.3 sessions are not established until after the main handshake has completed. The server sends a separate post-handshake message to the client containing the session details. Typically this will happen soon after the handshake has completed, but it could be sometime later (or not at all).

In some cases, it seems the server doesn't send the ticket unless the client sends a request.  This command just hangs there indefinitely:
```sh
openssl s_client -ign_eof -prexit -sess_out sess.pem -connect google.com:443 
```
But this one works as expected
```sh
echo -e "GET / HTTP/1.1\r\n" | openssl s_client -ign_eof -prexit -sess_out sess.pem -connect google.com:443

**Comment by Dridi** [comment]
I went ahead and spawned a new test case for this: cc5401f8a1be08d58f95b431cea1628aac2658cb.

This should help us get something more deterministic because currently the success rate of test35 is around 1% on two of my systems, until we take care of the TLS 1.3 special case.

**Comment by daghf** [comment]
Fixed via 0d6a6743f81473b0bfc8aa10ab8acd192c6ffbe0
